### DISCUSSION TITLE
Problems with Prisma client in a dockerized project

### DISCUSSION THREAD

**Discussion by modev** [prisma/prisma#8165] [Q&A] [answered]
Hey everyone,
I recently discovered Prisma and decided to give it a try during a test project, so far, I'm really loving it! However, I currently face a problem with Dockerizing my Backend. Hopefully I can get some help here :) 

My backend is a TypeScript + Express.js + Prisma project I configured manually, my package.json looks like that.
```json
{
  "name": "backend",
  "version": "1.0.0",
  "description": "",
  "main": "dist/index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1",
    "prebuild": "tslint -c tslint.json -p tsconfig.json --fix",
    "build": "tsc",
    "postinstall": "prisma generate",
    "start": "node .",
    "dev": "nodemon src/index.ts"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "dependencies": {
    "@prisma/client": "^2.26.0",
    "@types/body-parser": "^1.19.1",
    "@types/express": "^4.17.13",
    "@types/helmet": "^4.0.0",
    "@types/morgan": "^1.9.2",
    "@types/node": "^16.0.0",
    "body-parser": "^1.19.0",
    "express": "^4.17.1",
    "helmet": "^4.6.0",
    "morgan": "^1.10.0",
    "nodemon": "^2.0.9",
    "prisma": "^2.26.0",
    "ts-node": "^10.0.0",
    "tslint": "^6.1.3",
    "typescript": "^4.3.5"
  },
  "devDependencies": {
    "prettier": "^2.3.2"
  }
}
```
On my development machine, I'm using the Prisma client like that
```typescript
import express from "express";
import bodyParser from "body-parser";
import { register } from "./controllers";
import { PrismaClient } from ".prisma/client";

const app = express();
const prisma = new PrismaClient();

app.use(bodyParser.json());

app.get("/", (req, res) => res.send("Hello World!"));

register(app, prisma); // register application routers

app.listen(8070, () => console.log(`Listening on port 8070`));

export default app;

```
This setup works perfectly fine when running the nodemon `npm run dev` task on my machine. I created a Dockerfile with a two stage build process, but I always get the same error when initializing the container. 

```
Error: @prisma/client did not initialize yet. Please run "prisma generate" and try to import it again.
In case this error is unexpected for you, please report it in https://github.com/prisma/prisma/issues
    at new PrismaClient (/app/node_modules/.prisma/client/index.js:3:11)
    at Object.<anonymous> (/app/dist/index.js:11:16)
    at Module._compile (internal/modules/cjs/loader.js:1085:14)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1114:10)
    at Module.load (internal/modules/cjs/loader.js:950:32)
    at Function.Module._load (internal/modules/cjs/loader.js:790:14)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:76:12)
    at internal/main/run_main_module.js:17:47
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! backend@1.0.0 start: `node .`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the backend@1.0.0 start script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2021-07-08T13_32_43_998Z-debug.log
```

As I understand this message there is an issue with my generated client indicating that the client was not initialized although it should be. I initialized a post-install hook in the package.json which should generate the client. As I am executing the npm install command during the docker build stage and copy the whole node_modules folder to my prod stage, the Prisma client should be generated and be detected without problems. 

Am I missing something with my Dockerfile or my package.json configuration? I really appreciate any hints and solutions! 

```dockerfile
FROM node:14 AS build

WORKDIR /app
COPY package*.json .
RUN npm ci
COPY . . 
RUN npm run build

FROM node:14 AS prod
WORKDIR /app
COPY --from=build /app/dist dist/
COPY --from=build /app/node_modules node_modules/ 
COPY --from=build /app/prisma prisma/
COPY --from=build /app/package*.json .
EXPOSE 8070
CMD ["npm", "start"]
```
My directory structure looks something like this
```
dist/
src/
├─ controllers/
├─ index.ts
prisma/
node_modules/
Dockerfile
package.json

```

**Comment by ryands17** [comment, marked as answer]
@modev 👋 

I think running `npx prisma migrate` in the prod container should do the trick.

**Reply by modev** [reply]
Thank you, your answer pointed me into the right direction. I initially only copied the `package.json` and `package-lock.json` into my build stage and ran `npm install` on them. However, this resulted in the Prisma postinstall hook being executed while no .schema files were present in the container. Running `npx prisma generate` after copying the entirety of the project into the container it worked perfectly fine!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
